### DISCUSSION TITLE
Frontend testing framework(s)

### DISCUSSION THREAD

**Discussion by janiemi** [spraakbanken/korp-frontend#300] [General]
When visiting the [Protractor website](https://www.protractortest.org/), I noticed a note saying “Protractor is deprecated and will reach end-of-life in the summer of 2023”. Even without that, I suppose that moving away from [AngularJS](https://angularjs.org) will also mean that frontend testing will need to move to some other framework.

[Vue.js](https://vuejs.org) seems to recommend some [testing solutions](https://vuejs.org/guide/scaling-up/testing.html), such as [Cypress](https://www.cypress.io/) for [end-to-end testing](https://vuejs.org/guide/scaling-up/testing.html#e2e-testing) (with some [other options](https://vuejs.org/guide/scaling-up/testing.html#other-options-2)). However, I don’t know if they could be used at the transition stage from AngularJS to Vue.js.

**Comment by majsan** [comment]
It could be good to start writing new tests without Protractor before the Vue-rewrite, because the tests might help.

I have been looking a bit at different frontend testing libraries/frameworks. From what I saw, I did not like Cypress. To me, it looks a little heavy-weight. I prefer to run the tests as a simple CLI-application and it seems like Cypress does not support that (but maybe it does). If it seems much better than all alternatives than maybe that's what should be used anyway.

Here are the other alternatives I have found, but I don't have much information about them:
- Nightwatch
- testfront.io (Write tests by using the interface, the test framework records what is done) https://www.testfront.io/
- TestCafé https://testcafe.io/
- Taiko
- Pupeteer
- Playwright
- SeleniumBase (Python using pytest)
- WebdriverIO

I think some of these use Webdriver and some don't. One thing to check for each one is if whether Github Actions can run the tests.

**Reply by janiemi** [reply]
I agree that tests could help in avoiding bugs when converting to Vue.js.

I see that there are quite a few testing frameworks to choose from. I have no experience on any of the options myself, but [Vue.js testing page](https://vuejs.org/guide/scaling-up/testing.html#other-options-2) mentions [Playwright](https://playwright.dev/) and [Nightwatch v2](https://v2.nightwatchjs.org/) as the other options for end-to-end testing. Have I understood correctly that end-to-end testing is independent of the the application framework used?

I’m also wondering if it would also make sense to eventually write some [unit](https://vuejs.org/guide/scaling-up/testing.html#unit-testing) and/or [component tests](https://vuejs.org/guide/scaling-up/testing.html#component-testing). However, at least some aspects in them probably depend on the framework, so maybe it makes more sense after the conversion to Vue.js.

**Reply by majsan** [reply]
As far I now, yes, it should be possible to use any framework.

Component tests would be great. At least for the more complex components like the KWIC and extended search it would be helpful. Of course the tests could be on sub-components of those also.

**Comment by arildm** [comment]
I recently experimented a little with Playwright and it seems like a good choice. One advantage (in my opinion) is that it encourages accessible markup, e.g. with the `getByRole` locator which requires elements to have [WAI-ARIA roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles).

Temporary branch: https://github.com/spraakbanken/korp-frontend/tree/playwright

However, just like the Protractor test code, it is difficult to rely on it because API requests time out every now and then. I'm not sure if it is worth having flaky tests around for now or not...
